gocourse02/main.go

```diff
@@ -0,0 +1,134 @@
+package main
+
+import (
+	"fmt"
+	"math/rand/v2"
+)
+
+/*
+Написати програму “Зоопарк”. Звіри повтікали (більше трьох штук), наглядач повинен їх зібрати.
+Кожна сутність (наглядач, звір, клітка, …) представляється окремою структурою (zookeeper, animal, cage, …).
+Треба використати щонайменше: структури, вказівник, nil, будування, конструктор. Додати тваринам можливість
+розмножуватись. Програма має демонструвати свою роботу через вивід в stdout.
+*/
+
+const escapedAnimalsCount = 15 // Кількість звірів які повтікали
+//type Gender string
+//
+//const (
+//	male   Gender = "male"
+//	female Gender = "female"
+//)
+
+type ZooKeeper struct {
+	Name          string
+	AnimalsFound  int
+	AnimalsCaught int
+}
+
+func (k *ZooKeeper) SearchingForAnimalsReturnToCage(catch Catcher, animal *Animal, cage *Cage) {
+	if catch.FindEscapedAnimal() {
+		k.AnimalsFound++
+		if catch.AttemptCatchAnimal(cage, animal) {
+			k.AnimalsCaught++
+			cage.CurrentSeats++
+		}
+	}
+}
+
+type Catcher interface {
+	FindEscapedAnimal() bool                // Спроба знайти звіра
+	AttemptCatchAnimal(*Cage, *Animal) bool // Сброба зловити звіра в клітку
+}
+
+type Animal struct {
+	ID     int
+	Name   string
+	Gender string
+	Weight int
+}
+
+type Cage struct {
+	MaxSeats      int
+	MaxWeight     int
+	CurrentSeats  int
+	CurrentWeight int
+	Animal        []Animal
+}
+
+// Шукаємо звіра. Результат = Знаходимо або ні
+func (k *ZooKeeper) FindEscapedAnimal() bool {
+	return rand.N(2) == 1
+}
+
+// Намагаемося зловити в клітку знайденого звіра, якщо в клітці є місця  і не перевищує максимально дозволену вагу
+func (k *ZooKeeper) AttemptCatchAnimal(cage *Cage, a *Animal) bool {
+	if cage.MaxWeight >= a.Weight+cage.CurrentWeight && cage.MaxSeats >= cage.CurrentSeats+1 {
+		cage.CurrentWeight += a.Weight
+		cage.Animal = append(cage.Animal, *a)
+		return true
+	}
+	return false
+}
+
+// Функція розможноження
+func (male *Animal) Reproduction(female *Animal) {
+	fakeData := NewFake()
+	avgWeight := (male.Weight + female.Weight) / 2
+	NewAnimals := &Animal{ID: escapedAnimalsCount + 1, Name: fakeData.Name, Gender: fakeData.Gender, Weight: avgWeight}
```

Comment:
Дрібниця, але все ж:
```suggestion
	newAnimal := &Animal{ID: escapedAnimalsCount + 1, Name: fakeData.Name, Gender: fakeData.Gender, Weight: avgWeight}
```

Можна навіть перейменувати на `young` (чи `child`).

Reply:
Такий довгий рядок важко читається і модифікується, бо цілих чотири поля в структурі і всі в один рядок. А що буде, якщо в майбутньому додасться ще одне поле?

Пропоную ось таке форматування:

```suggestion
	child := &Animal{
		ID:     escapedAnimalsCount + 1,
		Name:   fakeData.Name,
		Gender: fakeData.Gender,
		Weight: avgWeight,
	}
```

Reply:
Спосіб фіксу 2: чи не можемо ми тут використати конструктор? Так, конструктор `NewAnimal` приймає лише параметр `id`. Але ж можна розширити конструктор.